### PR TITLE
Part 1: Make everything taggable Backend (starting with Actions) 

### PR DESCRIPTION
**Pull request by alexkim205**
## Please don't automerge until Part 3 and Part 2 make its way into this PR

**This is part of an effort to divide https://github.com/PostHog/posthog/pull/8331 up into more reviewable chunks.**

Part 2: https://github.com/PostHog/posthog/pull/8529
Part 3: https://github.com/PostHog/posthog/pull/8530

### PR dependency:

Part 1 <- Part 2 <- Part 3

## Changes

This PR lays the foundations for a new way of adding tags to any object. 

- [x] Adds `Tag` model
- [x] Adds `TaggedItem` model
- [x] Migrations for both
- [x] Add `TaggedItemSerializerMixin` for resolving viewset requests
- [x] Add `EnterpriseTaggedItemSerializerMixin` for resolving ee viewset requests
- [x] Make actions taggable and expose in frontend

## How did you test this code?

See https://github.com/PostHog/posthog/pull/8331

### CODE REVIEW

**Comment by alexkim205** [comment]
The last failing test affecting all three PRs seems to be unrelated to this task. Looks like it's from this https://github.com/PostHog/posthog/commit/2207bbe84e24a4641510c6a96cd1f0ec37e7accb. @Twixes

File: posthog/migrations/0206_global_tags_setup.py
Comment (mariusandra) [inline posthog/migrations/0206_global_tags_setup.py:63]: Total nit, but may be better to move `tag` to the top, as the new columns added at the end of the table will be similar to the other two. Looks weird to have `tag` sandwiched in the middle there

File: posthog/api/utils.py
Comment (mariusandra) [inline posthog/api/utils.py:291]: This seems rather magical. What's happening here under the hood?

File: posthog/models/tag.py
Comment (mariusandra) [inline posthog/models/tag.py:24]: Uhm, I think you [misunderstood my comment here](https://github.com/PostHog/posthog/pull/8331#issuecomment-1034332162). Tags should absolutely be able to have spaces inside. 

This line 

> The tag is only made up of letters, numbers, hyphens, and underscores

made it seem like it's not possible to have spaces in tags. If so, I think that's a mistake and we should have spaces.

File: posthog/migrations/0206_global_tags_setup.py
Comment (mariusandra) [inline posthog/migrations/0206_global_tags_setup.py:77]: I'm not convinced this makes sense at a database level. It feels like it'll just get needlessly complicated for little gain somehow 🤔

**Review by mariusandra — DISMISSED** [review]
Some thoughts inline. What this PR is missing is also a test for the new functionality. Feel free to include some part of part 3 to make that possible if it's complicated in isolation.

File: ee/api/ee_tagged_item.py
Comment (alexkim205) [reply ee/api/ee_tagged_item.py:13]: I put some thoughts here about the n+1 problem here https://github.com/PostHog/posthog/pull/8529#issuecomment-1034446986. I think I'll need to rewrite the mxin to intercept the request at the viewset level rather than the serializer level in order to use `prefetch_related`.

File: frontend/src/scenes/actions/actionEditLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/actions/actionEditLogic.ts:95]: Not a feature flag :)

File: frontend/src/scenes/actions/actionEditLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/actions/actionEditLogic.ts:104]: Why the lag?

File: posthog/migrations/0206_global_tags_setup.py
Comment (mariusandra) [inline posthog/migrations/0206_global_tags_setup.py:61]: We're not supporting tags on flags yet?

File: frontend/src/types.ts
Comment (mariusandra) [inline frontend/src/types.ts:1230]: not yet needed?

File: frontend/src/scenes/actions/actionEditLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/actions/actionEditLogic.ts:29]: We should not have these two actions at all.

File: frontend/src/scenes/actions/ActionEdit.tsx
Comment (mariusandra) [inline frontend/src/scenes/actions/ActionEdit.tsx:115]: Why are these two separate actions? Can't this use the traditional `value`/`onChange` pattern that we have on most controlled components?

**Review by mariusandra — DISMISSED** [review]
Alex, this project is honestly going on for way too long. The point of small PRs is that we build one complete part of the bigger feature, get that rock solid and bullet proof, and then merge it in. 

Right now I still can't give you a ✅ because of all the new code that has appeared. Even if some of it will be addressed by future PRs, I'm not going to look at them before this one is green. Neither should you.

This PR should ideally just add support for adding tags to any object. Including tests for the new functionality. I asked you to bring in some of the code for actions to have a shortcut for tests (no need to create mock "MyNewTaggyModel" models to add tags to, even if that's the cleaner approach I'd normally prefer), but you also brought in broken feature flag tagging code, a bunch of frontend code, and some typos.

Because of that, we'll have to do another review cycle, which is made worse by the timezones.

So, here's my ask, can you make one bullet proof PR that just does a few things, but does them well? The core of this PR is solid, but there's just now some mess around.

Said differently, 

1) Strip out all the frontend code
2) Strip out any mention of "feature flag", including prematurely added TS types, copy-pasted strings in the actions API, etc.

Just please get one solid thing that we can merge :).

I gave a bit of feedback for the frontend inline here, but it'll require more work. Ideally the frontend PR should also give us autocompletion, so you're not writing the same tag every time with the worry of making mistakes (`Tag` vs `tag`)...

File: posthog/migrations/0206_global_tags_setup.py
Comment (alexkim205) [reply posthog/migrations/0206_global_tags_setup.py:61]: Yes, this PR doesn't support FF's yet but I wanted to avoid having to run another migration for this in a future PR when we do.

Comment (alexkim205) [reply posthog/migrations/0206_global_tags_setup.py:61]: I can take this out if it will simplify things.

**Comment by alexkim205** [comment]
Hey Marius, appreciate your review on this. It looks like I misunderstood since I thought you wanted me to bring actions functionality into this PR as well, but instead it sounds like you wanted me to bring in just enough to write some tests for tagged items. 

The timezone difference makes it really hard to put out revisions reactively and ask for clarification here and ideally a lot of this back and forth could've been avoided with a few slack msgs. I'm not sure what the solution is here, but I'm a bit frustrated at the speed of this tagging situation as well just bc there's a lot of factors that are out of our control.

With that being said, I've stripped out everything FF related and removed all frontend code from this PR.

**Review by rcmarron — APPROVED** [review]
Looks good! 🎉

Let's hold off until Monday to merge it though!

**Comment by mariusandra** [comment]
This looks good! Thanks both!

@alexkim205 indeed the timezones are causing friction. The goal of the split was to always move fast, and I think that's what we both tried to do. However where I think we miscommunicated is that you split your original PR into 3 parts, and worked parts 1, 2 and 3 with the emphasis on getting it done at 3. However, you should have only created a new PR for part 1, and put all the effort into making that the tightest and smallest it could be.

I'm happy that @rcmarron also had time to feedback here. Without knowing how to guide you from the previous step to here myself, this does look like a nicer abstraction. 👍

**Pull request merged**
